# Введение в обработку текста на естественном языке

Материалы:
* Макрушин С.В. Лекция 9: Введение в обработку текста на естественном языке\
* https://realpython.com/nltk-nlp-python/
* https://scikit-learn.org/stable/modules/feature_extraction.html

## Задачи для совместного разбора

In [52]:
!pip3 install pymorphy2

In [53]:
from sklearn.feature_extraction.text import CountVectorizer
import pymorphy2

1. Считайте слова из файла `litw-win.txt` и запишите их в список `words`. В заданном предложении исправьте все опечатки, заменив слова с опечатками на ближайшие (в смысле расстояния Левенштейна) к ним слова из списка `words`. Считайте, что в слове есть опечатка, если данное слово не содержится в списке `words`.

In [54]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!


True

In [55]:
from nltk import edit_distance

with open('./data/litw-win.txt', 'r', encoding="utf-8") as f:
  txt = f.read().split("\n")
words = []
for i in txt:
  words.append(i.split()[-1])
words

['и',
 'в',
 'я',
 'с',
 'а',
 'к',
 'у',
 'о',
 'н',
 'п',
 'ж',
 'б',
 'т',
 'д',
 'м',
 'ч',
 'з',
 'г',
 'е',
 'р',
 'э',
 'л',
 'х',
 'ш',
 'ф',
 'ц',
 'щ',
 'й',
 'ю',
 'ы',
 'ъ',
 'ь',
 'не',
 'на',
 'он',
 'то',
 'но',
 'же',
 'вы',
 'по',
 'да',
 'за',
 'бы',
 'ты',
 'от',
 'из',
 'ее',
 'до',
 'ну',
 'ни',
 'ли',
 'уж',
 'во',
 'их',
 'мы',
 'со',
 'ей',
 'об',
 'ко',
 'ах',
 'им',
 'ка',
 'та',
 'пр',
 'те',
 'чт',
 'ту',
 'го',
 'де',
 'вс',
 'эт',
 'мо',
 'ра',
 'ст',
 'ха',
 'хе',
 'ум',
 'се',
 'эх',
 'гм',
 'ме',
 'св',
 'ль',
 'ег',
 'пе',
 'ва',
 'са',
 'ве',
 'ох',
 'бо',
 'хо',
 'че',
 'сл',
 'од',
 'бе',
 'мн',
 'ай',
 'ею',
 'ос',
 'ск',
 'ви',
 'ми',
 'ма',
 'сп',
 'ба',
 'ес',
 'бу',
 'гл',
 'що',
 'эй',
 'ем',
 'кр',
 'см',
 'др',
 'лю',
 'па',
 'ро',
 'зн',
 'ле',
 'як',
 'тр',
 'жи',
 'ел',
 'ус',
 'фу',
 'дв',
 'ру',
 'ал',
 'ой',
 'си',
 'ду',
 'вз',
 'ещ',
 'хи',
 'чи',
 'вд',
 'ил',
 'ух',
 'ча',
 'гр',
 'пи',
 'бл',
 'ре',
 'су',
 'ге',
 'ку',
 'оп',
 'ж

In [56]:
word = "велечайшим"
for w in words:
  if edit_distance(word, w)==1:
      print(w)

величайшим


In [57]:
import re

In [58]:
text = '''с велечайшим усилием выбравшись из потока убегающих людей Кутузов со свитой уменьшевшейся вдвое поехал на звуки выстрелов русских орудий'''

In [59]:
edit_distance("из",'ис')

1

In [60]:
splitText = re.split('\W+', text)
for index,s in  enumerate(splitText):
  for w in words:
    if edit_distance(s, w)==1 and len(s)>2:
      splitText[index] = w

print(splitText)

['с', 'величайшим', 'усилиям', 'выбрившись', 'из', 'потомка', 'убегающим', 'люден', 'кутузов', 'со', 'свиткой', 'уменьшившейся', 'вдвоем', 'поехало', 'на', 'звуко', 'выстрелив', 'прусских', 'орудуй']


2. Разбейте текст из формулировки задания 1 на слова; проведите стемминг и лемматизацию слов.

In [61]:
from nltk.stem import SnowballStemmer
from nltk import word_tokenize
import pymorphy2

In [62]:
text = '''Разбейте текст из формулировки второго задания на слова. Проведите стемминг и лемматизацию слов.'''

In [63]:
a = word_tokenize(word)
print(a)

['велечайшим']


In [64]:
stemmer = SnowballStemmer('russian')
for wordss in word_tokenize(text):
    print(stemmer.stem(wordss))

разб
текст
из
формулировк
втор
задан
на
слов
.
провед
стемминг
и
лемматизац
слов
.


In [65]:
morph = pymorphy2.MorphAnalyzer()
for wordss in word_tokenize(text):
    print(morph.parse(wordss)[0].normalized.word)

разбить
текст
из
формулировка
второй
задание
на
слово
.
провести
стемминг
и
лемматизация
слово
.


3. Преобразуйте предложения из формулировки задания 1 в векторы при помощи `CountVectorizer`.

In [66]:
from sklearn.feature_extraction.text import CountVectorizer
from nltk import sent_tokenize

In [67]:
text = '''Разбейте текст из формулировки второго задания на слова. Проведите стемминг и лемматизацию слов.'''

In [68]:
cv = CountVectorizer()
cv.fit(sent_tokenize(text))
sents_cv=cv.transform(sent_tokenize(text)).toarray()
sents_cv.shape, sents_cv

((2, 12),
 array([[1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1],
        [0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 0]]))

In [69]:
cv.vocabulary_

{'разбейте': 6,
 'текст': 10,
 'из': 2,
 'формулировки': 11,
 'второго': 0,
 'задания': 1,
 'на': 4,
 'слова': 8,
 'проведите': 5,
 'стемминг': 9,
 'лемматизацию': 3,
 'слов': 7}

## Лабораторная работа 9

In [70]:
import pandas as pd

### Расстояние редактирования

1.1 Загрузите предобработанные описания рецептов из файла `preprocessed_descriptions.csv`. Получите набор уникальных слов `words`, содержащихся в текстах описаний рецептов (воспользуйтесь `word_tokenize` из `nltk`).

In [71]:
data = pd.read_csv('./data/preprocessed_descriptions.csv', delimiter=',', parse_dates=["submitted"])

descriptions = data["description"].fillna("")
unique_words = []

for description in descriptions:
  tokens = word_tokenize(description)
  unique_words.extend(set(tokens))

# Преобразование списка уникальных слов в множество, чтобы удалить дубликаты, и обратно в список
unique_words = list(set(unique_words))

unique_words

['fry-pan',
 'crehan',
 'coal-burning',
 'thidabeau',
 'parragon',
 'trolly',
 'before',
 '135158',
 'bounty',
 'clonakilty',
 'omnivores',
 '87173',
 'lex',
 'rubbermaid',
 'geaux',
 'suzanne',
 'tarka',
 'frustrated',
 'grandson',
 "'finger",
 'pero',
 "'sticks",
 'sautéed',
 'bobbie',
 'best.',
 'induce',
 '141',
 'entrees',
 'minimize',
 'marlboro',
 'corners',
 'morsels',
 'tight-fitting',
 'syrp',
 'concoct',
 'bethenny',
 'foriegn',
 'unders',
 'sirop',
 'ahora',
 'coriander',
 'long-handled',
 'spices.the',
 'e.d',
 'sais',
 "'10",
 'exsisted',
 'tracks',
 'embosses',
 'following',
 'chile-',
 '255532',
 '1947.',
 '362382',
 'small-batch',
 'needing',
 '//www.foodlushblog.com/2011/03/recipe-knockoff-campbells-bean-with-bacon-soup.html',
 'clothes',
 "don't/",
 'tryed',
 'british/jewish',
 'overloaded',
 'welltellme.com',
 'activities',
 'bosley',
 'allow',
 '526666',
 'ran',
 'rossipasta.com',
 '1912',
 'nervous',
 'menthe',
 "'summer",
 'spouse',
 'cheesesticks',
 'wy',
 'suga

1.2 Сгенерируйте 5 пар случайно выбранных слов и посчитайте между ними расстояние редактирования.

In [72]:
import random

In [73]:
random_word_pairs = random.sample(unique_words, 10)

for i in range(0, len(random_word_pairs), 2):
  word1 = random_word_pairs[i]
  word2 = random_word_pairs[i+1]
  distance = edit_distance(word1, word2)
  print(f"Расстояние редактирования между '{word1}' и '{word2}': {distance}")

Расстояние редактирования между 'pureed' и 'taste-a-like': 11
Расстояние редактирования между 'this-but' и 'autism': 7
Расстояние редактирования между 'informations' и 'headed': 11
Расстояние редактирования между 'superbly' и 'justine': 7
Расстояние редактирования между 'penetrate' и 'www.thaifoodandtravel.com': 20


1.3 Напишите функцию, которая для заданного слова `word` возвращает `k` ближайших к нему слов из списка `words` (близость слов измеряется с помощью расстояния Левенштейна)

In [74]:
def find_k_nearest_words(word, words, k=5):
  sorted_words = sorted(words, key=lambda w: edit_distance(word, w))
  return sorted_words[:k]

word = "данные"
k_nearest = find_k_nearest_words(word, words)
k_nearest

['данные', 'данное', 'данных', 'данным', 'данный']

### Стемминг, лемматизация

2.1 На основе результатов 1.1 создайте `pd.DataFrame` со столбцами:
*   word
*   stemmed_word
*   normalized_word

Столбец `word` укажите в качестве индекса.

Для стемминга воспользуйтесь `SnowballStemmer`, для нормализации слов - `WordNetLemmatizer`. Сравните результаты стемминга и лемматизации.

In [75]:
from nltk.stem import WordNetLemmatizer
nltk.download('wordnet')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


True

In [76]:
stemmer = SnowballStemmer("english")
lemmatizer = WordNetLemmatizer()

processed_words = {
  'word': unique_words,
  'stemmed_word': [stemmer.stem(word) for word in unique_words],
  'normalized_word': [lemmatizer.lemmatize(word) for word in unique_words]
}

processed_words_df = pd.DataFrame(processed_words)
processed_words_df.set_index('word', inplace=True)

processed_words_df.head(10)

stemmed_word normalized_word
word                                     
fry-pan           fry-pan         fry-pan
crehan             crehan          crehan
coal-burning    coal-burn    coal-burning
thidabeau       thidabeau       thidabeau
parragon         parragon        parragon
trolly             trolli          trolly
before              befor          before
135158             135158          135158
bounty             bounti          bounty
clonakilty     clonakilti      clonakilty

2.2. Удалите стоп-слова из описаний рецептов. Какую долю об общего количества слов составляли стоп-слова? Сравните топ-10 самых часто употребляемых слов до и после удаления стоп-слов.

In [77]:
from nltk.corpus import stopwords
from collections import Counter
nltk.download('stopwords')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


True

In [78]:
stopwords = set(stopwords.words('english'))

filtered_words = [word for word in unique_words if word.lower() not in stopwords]

total_words = len(unique_words)
stopwords_count = total_words - len(filtered_words)
stopwords_percent = stopwords_count / total_words

original_top_10 = Counter(unique_words).most_common(10)
filtered_top_10 = Counter(filtered_words).most_common(10)

print(f"Доля стоп-слов: {stopwords_percent:.2%}")
print("Топ-10 слов до удаления стоп-слов:\n", original_top_10)
print("Топ-10 слов после удаления стоп-слов:\n", filtered_top_10)

Доля стоп-слов: 0.49%
Топ-10 слов до удаления стоп-слов:
 [('fry-pan', 1), ('crehan', 1), ('coal-burning', 1), ('thidabeau', 1), ('parragon', 1), ('trolly', 1), ('before', 1), ('135158', 1), ('bounty', 1), ('clonakilty', 1)]
Топ-10 слов после удаления стоп-слов:
 [('fry-pan', 1), ('crehan', 1), ('coal-burning', 1), ('thidabeau', 1), ('parragon', 1), ('trolly', 1), ('135158', 1), ('bounty', 1), ('clonakilty', 1), ('omnivores', 1)]


### Векторное представление текста

3.1 Выберите случайным образом 5 рецептов из набора данных. Представьте описание каждого рецепта в виде числового вектора при помощи `TfidfVectorizer`

In [79]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [82]:
random_recipes = data.sample(5)
random_recipes_descriptions = random_recipes['description'].astype(str).tolist()

vectorizer = TfidfVectorizer()

tfidf_matrix = vectorizer.fit_transform(random_recipes_descriptions)

feature_names = vectorizer.get_feature_names_out()
tfidf_vectors = tfidf_matrix.toarray()

for i, desc in enumerate(random_recipes_descriptions):
  print(f"Рецепт {i + 1}:")
  print(desc)
  print(f"\nЧисловой вектор:\n{tfidf_vectors[i]}\n")

Рецепт 1:
tuna with a blast of wasabi. we nearly always have this with sliced crisp-tender green beans folded into the noodles. finely sliced snow peas work well too.

Числовой вектор:
[0.         0.         0.18188717 0.         0.         0.18188717
 0.         0.18188717 0.         0.         0.         0.
 0.         0.18188717 0.18188717 0.18188717 0.         0.
 0.         0.         0.         0.         0.18188717 0.18188717
 0.         0.         0.         0.         0.         0.18188717
 0.         0.         0.         0.         0.18188717 0.
 0.18188717 0.18188717 0.         0.         0.         0.
 0.18188717 0.         0.         0.         0.         0.
 0.36377434 0.18188717 0.         0.         0.         0.18188717
 0.18188717 0.         0.08667021 0.18188717 0.         0.18188717
 0.         0.18188717 0.18188717 0.18188717 0.         0.36377434
 0.18188717 0.         0.         0.        ]

Рецепт 2:
this one is a bit healthier and there is no cream soup in it.

3.2 Вычислите близость между каждой парой рецептов, выбранных в задании 3.1, используя косинусное расстояние (`scipy.spatial.distance.cosine`) Результаты оформите в виде таблицы `pd.DataFrame`. В качестве названий строк и столбцов используйте названия рецептов.

In [83]:
from scipy.spatial.distance import cosine

In [86]:
random_recipes

name      id  minutes  \
29057  wasabi tuna steaks with hokkien or udon noodles  178540       70   
5508                    chicken   brown rice casserole  192281       85   
29947                   zucchini meat sauce with pasta  253264       30   
26597                        sweet and sour bean salad  187954       25   
10775                            fassoulada  bean soup  424172      160   

       contributor_id  submitted  n_steps  \
29057          279205 2006-07-20      9.0   
5508           366730 2006-10-25      NaN   
29947          275742 2007-09-17      NaN   
26597           37449 2006-09-27     13.0   
10775          178427 2010-05-10      NaN   

                                             description  n_ingredients  \
29057  tuna with a blast of wasabi. we nearly always ...            6.0   
5508   this one is a bit healthier and there is no cr...            9.0   
29947  havn't tried this recipe yet but hopefully wil...            6.0   
26597  a great recipe from paula's home cooking, you'...           14.0   
10775  i found this on alleasyrecipes.com. i'm postin...            NaN   

                                preprocessed_description  
29057  tuna with a blast of wasabi we nearly always h...  
5508   this one is a bit healthier and there is no cr...  
29947  havnt tried this recipe yet but hopefully will...  
26597  a great recipe from paulas home cooking youll ...  
10775  i found this on alleasyrecipescom im posting i...

In [94]:
recipe_names = random_recipes['name'].tolist()

tfidf_matrix_array = tfidf_matrix.toarray()

distances = pd.DataFrame(index=recipe_names, columns=recipe_names)

for i in range(len(recipe_names)):
  for j in range(len(recipe_names)):
    if i != j:
      distances.iloc[i, j] = cosine(tfidf_matrix_array[i], tfidf_matrix_array[j])
    else:
      distances.iloc[i, j] = 0.0  # Расстояние между рецептом и самим собой

distances = distances.astype(float)

distances

wasabi tuna steaks with hokkien or udon noodles  \
wasabi tuna steaks with hokkien or udon noodles                                         0.000000   
chicken   brown rice casserole                                                          0.992165   
zucchini meat sauce with pasta                                                          0.984949   
sweet and sour bean salad                                                               0.989276   
fassoulada  bean soup                                                                   0.985859   

                                                 chicken   brown rice casserole  \
wasabi tuna steaks with hokkien or udon noodles                        0.992165   
chicken   brown rice casserole                                         0.000000   
zucchini meat sauce with pasta                                         0.939295   
sweet and sour bean salad                                              0.924681   
fassoulada  bean soup                                                  0.900680   

                                                 zucchini meat sauce with pasta  \
wasabi tuna steaks with hokkien or udon noodles                        0.984949   
chicken   brown rice casserole                                         0.939295   
zucchini meat sauce with pasta                                         0.000000   
sweet and sour bean salad                                              0.916913   
fassoulada  bean soup                                                  0.971665   

                                                 sweet and sour bean salad  \
wasabi tuna steaks with hokkien or udon noodles                   0.989276   
chicken   brown rice casserole                                    0.924681   
zucchini meat sauce with pasta                                    0.916913   
sweet and sour bean salad                                         0.000000   
fassoulada  bean soup                                             0.921935   

                                                 fassoulada  bean soup  
wasabi tuna steaks with hokkien or udon noodles               0.985859  
chicken   brown rice casserole                                0.900680  
zucchini meat sauce with pasta                                0.971665  
sweet and sour bean salad                                     0.921935  
fassoulada  bean soup                                         0.000000

3.3 Какие рецепты являются наиболее похожими? Прокомментируйте результат (словами).

Все рецепты в большей мере отличаются друг от друга, но есть два, которые немного, но похожи, если судить по их косинусному расстоянию - это chicken brown rice casserole и fassoulada  bean soup. У них косинусное расстояние 0.900680, что является самым малым, не считая рецептов, которые сравниваются сами с собой.